storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+	// with a digest-based name that can't be used as a map key.
+	ErrInvalidBlobDigest = errors.New("invalid blob digest")
+	// ErrBlobDigestMismatch is returned when PutBlob() is given a blob
+	// with a digest-based name that doesn't match its contents.
+	ErrBlobDigestMismatch = errors.New("blob digest mismatch")
+	// ErrManifestDigestMismatch is returned when GetTargetManifest() is
+	// given a digest-based name that doesn't match the manifest.
+	ErrManifestDigestMismatch = errors.New("manifest digest mismatch")
+	// ErrNoSuchImage is returned when we attempt to access an image which
+	// doesn't exist in the storage area.
+	ErrNoSuchImage = storage.ErrNotAnImage
+)
+
+type storageImage struct {
+	store          storage.Store
+	imageRef       *storageReference
+	Generic        types.Image
+	Tag            string            `json:"tag,omitempty"`
+	Created        time.Time         `json:"created-time,omitempty"`
+	ID             string            `json:"id"`
+	BlobList       []string          `json:"blob-list,omitempty"` // Ordered list of every blob the image has been told to handle
+	Layers         map[string]string `json:"layers,omitempty"`    // Map from names of blobs that are layers to layer IDs
+	BlobData       map[string][]byte `json:"-"`                   // Map from names of blobs that aren't layers to contents, temporary
+	SignatureSizes []int             `json:"signature-sizes"`     // List of sizes of each signature slice
+}
+
+type storageLayerMetadata struct {
+	ExpectedSize int64  `json:"expected-size,omitempty"`
+	Digest       string `json:"digest,omitempty"`
+	Size         int64  `json:"size,omitempty"`
+}
+
+// newImageSource sets us up to read out an image, which we assume exists.
+func newImageSource(imageRef *storageReference) (*storageImage, error) {
+	tag := ""
+	if imageRef.ID() == "" {
+		logrus.Errorf("no image matching reference %q found", imageRef.StringWithinTransport())
+		return nil, ErrNoSuchImage
+	}
+	img, err := imageRef.store.GetImage(imageRef.ID())
+	if err != nil {
+		logrus.Errorf("error reading image %q: %v", imageRef.ID(), err)
+		return nil, err
+	}
+	tag = imageRef.StringWithinTransport()
+	image := storageImage{
+		store:    imageRef.store,
+		imageRef: imageRef,
+		Tag:      tag,
+		Created:  time.Now(),
+		ID:       img.ID,
+		BlobList: []string{},
+		Layers:   make(map[string]string),
+		BlobData: make(map[string][]byte),
+	}
+	if err := image.loadMetadata(); err != nil {
+		logrus.Errorf("error decoding metadata for source image: %v", err)
+		return nil, err
+	}
+	return &image, nil
+}
+
+// newImageDestination sets us up to write a new image.
+func newImageDestination(imageRef *storageReference) (*storageImage, error) {
+	// We set the image's ID if the reference we got looked like one, since
+	// we take that as an indication that it's going to end up with the
+	// same contents as an image we already have.  If the reference looks
```

Comment:
It was sort of chaotic, since Commit() wasn't doing as much heavy lifting as it could because I didn't move much into it.  Reworking that so that the semantics are that: if an ID is supplied and it conflicts with one we already have in the store, we'll return an error.  If a name is supplied and it duplicates one that we already have in the store, the name will be assigned to point to the image we're writing.